sparkly/testing.py

```diff
@@ -120,28 +147,132 @@ def tearDown(self):
+
+        Extend this classic function signature to work better with
+        comparisons involving dataframes, rows, dictionaries and lists by
+        ignoring the order of mutable nested structures recursively and
+        rounding up floats when necessary.
+
+        Args:
+            first, second, msg: see unittest.TestCase.assertCountEqual
+            ignore_order (bool|False): ignore the order in all structures
+                that are inherently oredered (e.g., lists, mock calls).
+            max_recursion_depth (int|None): if ignore_order is true, do
+                ignore order up to this nesting level (exclusive). Setting
+                this to 0 is same to setting ignore_order=False, 1 means
+                ignore order only at the top level and so on. Default is
+                ignore order arbitrarily deep.
+            places (int|None): Decimal or significant digits to consider
+                during comparisons of floats.
+            significant (bool|False): If set, interpret places as # of
+                significant digits instead of decimal.
+            equal_nan (bool|True): If set, NaNs will compare equal.
+
+        Returns:
+            None, iff the two objects are equal.
+
+        Raises
+            AssertionError, iff the two objects are not equal. See
+            unittest.TestCase.assertEqual for details.
+        """
+        def _create_key(entry):
+            if isinstance(entry, type(mock.call)):
+                _, call_args, call_kwargs = entry
+                entry = list(call_args) + list(call_kwargs.items())
+            if isinstance(entry, collections.Iterable) and not isinstance(entry, six.string_types):
+                return entry is None, tuple(_create_key(x) for x in entry)
+            return entry is None, entry
+
+        def _nicefy(obj, recursion_depth=0):
+            if isinstance(obj, float):
+                if math.isnan(obj):
+                    # assertEqual does not recognize two nans as being
+                    # equal otherwise
+                    if equal_nan:
+                        obj = 'nan'
+                elif places is not None:
+                    if significant:
+                        obj = float('%.{}g'.format(places) % obj)
+                    else:
+                        obj = round(obj, places)
+
+            if isinstance(obj, Row):
+                obj = obj.asDict(recursive=True)
+
+            if isinstance(obj, T.DataType):
+                obj = obj.jsonValue()
+
+            if isinstance(obj, type(mock.call)):
+                try:
+                    _, call_args, call_kwargs = obj
+                except ValueError:  # not called from a class instance
+                    call_args, call_kwargs = obj
+
+                obj = mock.call(
+                    *[_nicefy(call_arg, recursion_depth + 1) for call_arg in call_args],
+                    **collections.OrderedDict(_nicefy(call_kwargs, recursion_depth + 1))
```

Comment:
Yeah, it was a hack to make `_create_key` work. Got rid of support for `mock.calls` since they're not related to Spark in any way, and it's easy to build it when needed on top of the new `assertEqual`